File: src/main/java/xyz/dream/launcher/managers/Folderer.java
// Copyright 2025 © Arsenii Motorin
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package xyz.dream.launcher.managers;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

public class Folderer {

    /**
     * Creates a folder for Minecraft files.
     * The folder is created in the user's home directory or the appropriate application data directory.
     *
     * @return The path to the created folder as a String.
     */
    public static String createTheFolder() {
        String os = System.getProperty("os.name").toLowerCase();
        String userHome = System.getProperty("user.home");
        String folderPath;
        
        if (os.contains("mac")) {
            folderPath = userHome + "/Library/Application Support/DreamLauncher";
        } else if (os.contains("win")) {
            folderPath = System.getenv("APPDATA") + "\\DreamLauncher";
        } else {
            folderPath = userHome + "/DreamLauncher";
        }
        
        try {
            Path path = Paths.get(folderPath);
            if (!Files.exists(path)) {
                Files.createDirectories(path);
                System.out.println("DreamLauncher folder created at: " + folderPath);
            }
        } catch (Exception e) {
            System.err.println("Error creating DreamLauncher directory: " + e.getMessage());
        }
        
        return folderPath;
    }
}

File: src/managers/folderer.rs
// Copyright 2025 © Arsenii Motorin
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::fs;

// Function that returns the path to the DreamLauncher folder based on the operating system
pub fn folderer() -> String {
    // cfg! is a Rust attribute that allows conditional compilation based on the target operating system
    // It checks if the target OS is macOS, Windows, or any other OS and returns
    // cfg! returns a boolean value indicating whether the condition is true or false
    let folder_path = if cfg!(target_os = "macos") {

        // unwrap() is used to get the value inside an option or result
        // If the value is None or Err, it will panic
        dirs::home_dir().unwrap().join("Library/Application Support/DreamLauncher")
    } else if cfg!(target_os = "windows") {
        dirs::data_dir().unwrap().join("DreamLauncher")
    } else {
        dirs::home_dir().unwrap().join("DreamLauncher")
    };

    // Error handling for creating the directory
    if let Err(e) = fs::create_dir_all(&folder_path) {
        eprintln!("Error creating DreamLauncher directory: {}", e);
        } else {
        println!("DreamLauncher folder created at: {}", folder_path.display());
    }
    
    // Return
    folder_path.to_string_lossy().to_string()
}